netbird: rework server and include new component

PatrickDaG opened the pull request:
This is a pretty hefty rework of the nixos netbird modules.

First of all I split the package into three because currently you cannot have the client installed without the server components coming with it, now it's three packages, a client, a client with gui and a server.
You still have the option to build a package containing everything but I don't think most people need that.

Secondly I wrote a basic test for the server, now we at least know if it starts, which it currently doesn't cause upstream introduced clashing ports for all server, that cannot be disabled.
I would love further testing but I think that would need actually logging in into the kanidm instance inside the testing framework, which is something for another day.
The test also currently depend on #353681.

Netbird is currently switching away from coturn in favour of their own relay implementation, which this pull adds.
Their communication towards whether coturn will be needed going forward is a bit confusing, but I'm pretty sure right now you need both their relay and coturn, maybe in a few updates we can remove coturn.

Lastly I reworked the nginx setup, realizing you don't necesarrily need it, apart from serving the dashboard.
I removed it from all services and the default setup should now work without it, but you have to forward and open all relevant ports, for the management, signal, coturn, dashboard and relay.

To make it easier for people using nginx as a reverse proxy I've added the proxy, module which is written and maintained completely by myself and has no affiliation to upstream netbird. I do plan on using this and think it's valuable to have even just as a documentation of nginx options to use with netbird, but I am scared that people have problems with this module and complain to upstream netbird. Don't really know what to do whether to include or not, feedback appreciated.

In general this isn't extensively tested yet, but I would be very happy if people help me test it.
It has to wait for branch off anyway because it contains a bunch of breaking changes.

Also should probably write more documentation especially regarding the proxy module.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Saturn745 commented on nixos/modules/services/networking/netbird/dashboard.nix:
```diff
@@ -39,8 +39,6 @@ in
 
     package = mkPackageOption pkgs "netbird-dashboard" { };
 
-    enableNginx = mkEnableOption "Nginx reverse-proxy to serve the dashboard";
```
Should `mkRemovedOptionModule` be added for thhis?

PatrickDaG replied:
I've thought about it a bit more and realized that people might want to host their own dashboard, without nginx, but using the `finalDRV` so we might need the option after all.

Saturn745 commented:
Result of `nixpkgs-review pr 354032` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>netbird</li>
    <li>netbird-server</li>
    <li>netbird-ui</li>
  </ul>
</details>

vrifox commented:
> Result of `nixpkgs-review pr 354032` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 2 packages blacklisted:
> nixos-install-tools tests.nixos-functions.nixos-test
> 
> 3 packages built:
> netbird netbird-server netbird-ui

Same here, also on x86_64-linux. :)

TheRealGramdalf commented:
Fellow user of Kanidm/Netbird here. Can confirm that the server is non functional at the moment, again due to clashing ports. For now I'm going to have to override the signal/mgmt metrics port manually (by editing the systemd service), since there's currently no extraArgs setting or the like for the netbird components. Is it worth creating a separate PR to fix that in the meantime?

Also please let me know if I can help with testing in any way.

PatrickDaG commented:
Might be a good idea to put the `extraOption` in separate PR so they get merged sooner. I feel like this one still has quite the journey ahead.
You wanna do it? Else I can make one in a few days.

More eyes and people testing are always welcome. I'm just gonna run the new module while I work on this PR for a few weeks and see if anything comes up. Seems like right now the blocker for both of us is kanidm https://github.com/kanidm/kanidm/issues/3217

TheRealGramdalf commented:
> Might be a good idea to put the `extraOption` in separate PR so they get merged sooner. I feel like this one still has quite the journey ahead.
> You wanna do it? Else I can make one in a few days.

Yeah, can do - I was thinking add an extraOption to signal, and potentially even adding my snippet above as the default for one of them so it works out of the box again?
Or would adding a dedicated metrics option be better?

PatrickDaG commented:
I would say you could just add your snippet as the default and we can always switch to a dedicated option later, but thinking about it I'm not sure the newest update will work without the relay server packaged and setup so I'm not sure a new PR is worth it if it won't work anyway.

TheRealGramdalf commented:
> ...but thinking about it I'm not sure the newest update will work without the relay server packaged and setup so I'm not sure a new PR is worth it if it won't work anyway.

I'm on nixpkgs `dc460ec76cbff0e66e269457d7b728432263166c`, netbird v0.31.0, and my peers using setup keys (i.e. not depending on OIDC) are functioning just fine. #356512 was just merged though so I'm not sure if that changes things, the release notes don't seem to indicate anything however. I'll open a PR to add the snippet and hold off on a dedicated option for now, and I'll throw in an `extraOptions` to signal to make things easier in case things change down the line.

PatrickDaG commented:
> I'm on nixpkgs `dc460ec76cbff0e66e269457d7b728432263166c`, netbird v0.31.0, and my peers using setup keys (i.e. not depending on OIDC) are functioning just fine.

Oh nice. Wasn't sure how much they already depend on the relay.

> I'll open a PR to add the snippet and hold off on a dedicated option for now, and I'll throw in an `extraOptions` to signal to make things easier in case things change down the line.

Sounds great.

I'll try and finish this one soon as well, now that the branchoff happened.

TheRealGramdalf commented:
PR opened, input would be appreciated @PatrickDaG

f0x52 commented on nixos/modules/services/networking/netbird/server.nix:
```diff
@@ -53,31 +52,35 @@ in
+      };
+
+      relay = {
+        settings.NB_EXPOSED_ADDRESS = mkDefault "rel://${cfg.domain}/${builtins.toString cfg.relay.port}";
```
there's a typo in this default, the separator should be `:` (`${cfg.domain}:${...}`) instead of a `/`

ser commented:
I would like motivate you to finalise this PR, as the new relay function is very important for all self-hosted installations...

f0x52 commented:
Agreed, I've been using the patches from this PR for the last month now (with the small fix I reviewed), and it's worked really well on a combination of aarch64 and x86 systems.

Without the relay server / using the old module I could not get netbird to connect peers reliably at all

PatrickDaG committed:
netbird: refactor module

PatrickDaG committed:
nixos/netbird: add server tests

nixos/netbird: introduce standalone proxy module

PatrickDaG committed:
nixos/netbird: introduce netbirds native relay

PatrickDaG commented:
I think it's ready for review. Let me know if I missed anything.

ser commented:
@fricklerhandwerk @GetPsyched @hsjobeki @Saturn745 

Hi, I do understand that review of this code is not simple, but I would like to encourage you to make the effort as it's last missing step to make the package available. The new relay function is crucial for having a working netbird instance so the previous version of software is mostly useless at the moment.

fricklerhandwerk commented on nixos/modules/services/networking/netbird/server.md:
````diff
@@ -4,39 +4,57 @@ NetBird is a VPN built on top of WireGuard® making it easy to create secure pri
+The proxy assumes being hosted over https.
+
+When using the proxy remember to set all public address options to use the proxy host instead of the instance itself. These include:
+```nix
+server.domain
+management.settings.Signal.URI
+relay.setting.NB_EXPOSED_ADDRESS
+
+```
````
This should rather be a list since its not a nix expression:

```suggestion
- `server.domain`
- `management.settings.Signal.URI`
- `relay.setting.NB_EXPOSED_ADDRESS`
```

fricklerhandwerk commented on nixos/modules/services/networking/netbird/server.md:
```diff
@@ -4,39 +4,57 @@ NetBird is a VPN built on top of WireGuard® making it easy to create secure pri
+## Proxy {#module-services-netbird-server-proxy}
+The proxy module sets up a unified nginx proxy in front of your netbird instance.
+
+> [!CAUTION]
+> The proxy module is completely independent from netbird upstream. If you have any problems with it
+> DO NOT open an issue in any `netbirdio` repositories. Instead open them in `nixos/nixpkgs` and ping the
+> proxy maintainers.
```
```suggestion
:::{.warning}
The proxy module is completely independent from netbird upstream.
If you have any problems with it DO NOT open an issue in any `netbirdio` repositories. Instead open them in [`nixos/nixpkgs`](https://github.com/nixos/nixpkgs/issues) and ping the proxy maintainers.
:::
```

We have our own custom admonitions, not GitHub markdown. Please check the Nixpkgs manual for details, I may have remembered the syntax wrong.

fricklerhandwerk commented on nixos/modules/services/networking/netbird/server.md:
```diff
@@ -4,39 +4,57 @@ NetBird is a VPN built on top of WireGuard® making it easy to create secure pri
+A list of supported SSOs and their setup are available [on Netbird's documentation](https://docs.netbird.io/selfhosted/selfhosted-guide#step-3-configure-identity-provider-idp).
 
-There are quite a few settings that need to be passed to Netbird for it to function, and a minimal config looks like :
+There are quite a few settings that need to be passed to Netbird for it to function, and a minimal config might look like :
```
Please use an `:::{.example}` here, that way it will show up in the index of examples.

fricklerhandwerk commented on nixos/modules/services/networking/netbird/proxy.nix:
```diff
@@ -0,0 +1,106 @@
+
+    signalAddress = mkOption {
+      type = str;
+      description = "The external address to reach the signal service.";
```
What's that address? If it's IP, we have types for that which are stricter than just string.

Assuming it is, I'd reflect that in the wording:

```suggestion
      description = "External IP address to reach the signal service.";
```

(Prefixing "The" to all entries adds no information, and I suggest leaving that out.)

fricklerhandwerk commented on nixos/modules/services/networking/netbird/proxy.nix:
```diff
@@ -0,0 +1,106 @@
+
+    domain = mkOption {
+      type = str;
+      description = "The public domain to reach the proxy";
```
```suggestion
      description = "Public DNS domain name to reach the proxy";
```

"Public domain" triggers association with licensing, and I suggest avoiding distraction to readers even if it's ever so slight.

fricklerhandwerk commented on nixos/modules/services/networking/netbird/server.md:
````diff
@@ -4,39 +4,57 @@ NetBird is a VPN built on top of WireGuard® making it easy to create secure pri
+There are quite a few settings that need to be passed to Netbird for it to function, and a minimal config might look like :
 
 ```nix
 services.netbird.server = {
````
It wasn't there before, but since you're touching it: please make examples self-contained. In this case it should be a complete module expression. Eventually we want examples to be tested in CI, and until then we can slowly move them to be testable in the first place.

fricklerhandwerk reviewed (commented):
A few documentation nits, looks good overall; didn't review the code (I get auto-assigned as reviewer when something about the sections in the manual changes).

PatrickDaG commented:
Should have addressed the feedback. Thx.

ser commented:
@GetPsyched @hsjobeki @Saturn745 @vrifox 

Hello guys, could you spend a while to help closing this PR, as it's very important for VPN service owners. Netbird in current version is failing more and more.

GetPsyched commented:
Is there anything specific you want me to review? I'm not sure about netbird, so I can't be of much help there.

LoCrealloc reviewed (commented):
This is just a small code review for some things I noticed. I yet have to test the changes. I am a bit worried though that the new proxy module might be a bit too opinionated, but I will have to look into this a more.

Defelo commented on nixos/modules/services/networking/netbird/proxy.nix:
```diff
@@ -0,0 +1,110 @@
+{
+  options.services.netbird.server.proxy = {
+
+    enableNginx = mkEnableOption "Nginx reverse-proxy for the netbird signal service";
```
`mkEnableOption` already adds the "Whether to enable" part (see https://github.com/NixOS/nixpkgs/blob/master/lib/options.nix#L187)

LoCrealloc replied:
Alright, thanks. Still, the current description focusses on the signal module, though the proxy module adds proxy functionality for all netbird components.

PatrickDaG replied:
Yeah. Seems I forgot to change it after copying from the signal options. Is the new description better?

PatrickDaG commented on nixos/modules/services/networking/netbird/management.nix:
```diff
@@ -58,10 +58,14 @@ let
 
     Signal = {
       Proto = "https";
-      URI = "${cfg.domain}:443";
```
This was just copied from the upstream definitions. If you enable the whole stack this will get set in `server.nix`. If you don't enable the whole stack I don't think we can set a sane default there.

PatrickDaG commented on nixos/modules/services/networking/netbird/server.nix:
```diff
@@ -53,13 +52,13 @@ in
+          turnPort = config.services.coturn.listening-port;
           # We cannot merge a list of attrsets so we have to redefine the whole list
           settings = {
+            Signal.URI = mkDefault "${cfg.domain}:${builtins.toString cfg.signal.port}";
```
Nothing. Should be fixed by moving into its own `optionalAttrs` block.

PatrickDaG commented on nixos/modules/services/networking/netbird/management.nix:
```diff
@@ -220,95 +224,93 @@ in
-          Datadir = "''${stateDir}/data";
-          DataStoreEncryptionKey = "genEVP6j/Yp2EeVujm0zgqXrRos29dQkpvX0hHdEUlQ=";
-          StoreConfig = { Engine = "sqlite"; };
+        Datadir = "/var/lib/netbird-mgmt/data";
```
just an oversight. Should be fixed.

PatrickDaG commented:
I think I've addressed the feedback. Let me know if I missed something.

Regarding the proxy module, I can totally understand if you think it's too opinionated, or should not be included in nixpkgs. I just thought it be a good idea to have a singular exposed frontend, based mainly on [this file](https://github.com/netbirdio/netbird/blob/main/infrastructure_files/nginx.tmpl.conf) from netbird upstream. Netbirds seems to need some specific forwarding for all the services and I wanted to save others from having to go looking around the netbird codebase to find them.

PatrickDaG commented on nixos/modules/services/networking/netbird/management.nix:
```diff
@@ -128,18 +132,18 @@ let
   managementFile = settingsFormat.generate "config.json" managementConfig;
 
-  cfg = config.services.netbird.server.management;
+  cfg = config.services.netbird.server;
```
Oh sorry. I think missunderstood you're comment. I was using `cfg.relay` at two locations. But I just changed that just now because it seemed like a bad idea.

jvanbruegge commented:
I currently have issues with the netbird server components (clients loosing connection to the server), so I will give this PR a spin over the weekend.
Maybe you also want to add a commit to update netbird, the current version is 0.43.1.

jvanbruegge commented:
Oh, master already has the newest version, so maybe a rebase is in order.

jvanbruegge commented:
That was fast :sweat_smile:

jvanbruegge commented:
Ok, I've set it up now and it works, but I have two comments:

1. There should be an easy way to use the same nginx to host the dashboard and for the proxy module. Currently this is prevented by the proxy module wanting to forward to another address. It could be fixed by making the dashboard proxyPass conditional on `cfg.domain != config.services.netbird.dashboard.domain`. I used this as a workaround:
```nix
services.nginx.virtualHosts."${NETBIRD_DOMAIN}" = {
  locations."/" = lib.mkForce {
    root = config.services.netbird.server.dashboard.finalDrv;
    tryFiles = "$uri $uri.html $uri/ =404";
  };
};
```

2. The dashboard is a bit out of date, the newest version is 2.11.0

Let's see if this fixes my issues with the vpn disconnecting and not reconnecting

PatrickDaG commented:
Good points, thanks. I've added an option to disable proxying of the dashboard and opened a new pull requests to update the dashboard.

jvanbruegge commented:
Had this running for a week now and it works very well. Either this or the update to 0.43.1 fixed the issues I had with the clients disconnecting and the dashboard showing no clients (even if the underlying wireguard tunnel was still active)

PatrickDaG commented:
Just as an update. I don't use netbird anymore, and currently don't intend to switch back to it anytime soon, so my motivation to keep pushing this PR is limited. If anyone wants to adopt it I'd be more than happy to hand it over. I think it should be basically merge ready for now, so the only thing missing is someone willing to pester committers to get it over the line.

RafaelKr commented:
> I'm not terribly happy with the multi-user situation in netbird right now, most features are only available on their cloud version.

@TheRealGramdalf Could you elaborate on what you mean with the multi-user situation? We're currently evaluating using netbird in an enterprise environment.

BTW with version 0.64.0 (released yesterday) coturn is not required anymore: https://github.com/netbirdio/netbird/releases/tag/v0.64.0

schromp commented:
i have resolved the merge conflicts aswell as tested 0.63.0 yesterday. (going to do 0.64 today i guess)
Any way i can have access to this PR to update it or should i create a new one?

Btw ive been running netbird based on this PR for about 5 months now

schromp commented:
> Is there any update on this PR or a new one? @schromp 

No i did not have time sadly.

https://github.com/NixOS/nixpkgs/pull/487367 got opened though

woile commented:
hey @PatrickDaG is it okay if I start new PR's based on your work? Since you are not using netbird anymore, I want to slowly start pushing one module at a time (starting with the relay). 

I have to run everything on my configuration first to make sure it works. Any advice I should keep in mind?

PatrickDaG commented:
Absolutely, go for it. You might also want to look at #487367 and the people who commented, so you don't get in each others way.
I'll close this PR since I am pretty sure I will not come back to it, but if you or anyone can reuse parts, I'd be happy to know the work wasn't completely wasted.

It's been so long since I've run this stack, I don't remember much, sorry.
